kernel/mmanutils.c

```diff
@@ -538,12 +561,31 @@ static int _remove_file_mappings(void* addr, size_t length)
     _rlock(&locked);
     {
         const size_t count = length / PAGE_SIZE;
+        int prev_cleared_fd = -1;
 
         for (size_t i = index; i < index + count; i++)
         {
-            myst_fdmapping_t* p = &v.fdmappings[index];
-
             /* remove any fd-mapping (it is okay if it does exist) */
+            myst_fdmapping_t* p = &v.fdmappings[i];
+
+            /* add fd's for in-use mappings to a list. add only for the first
+             * page of
+             * interval mapped with same fd. The fd list will be closed by the
+             * caller outside the mman lock. */
+            if (p->used == MYST_FDMAPPING_USED && p->fd != prev_cleared_fd)
+            {
+                fdlist_t* fd_node = calloc(1, sizeof(fdlist_t));
```

Comment:
possible memory allocation failure Make sure the unlock happen!

Reply:
Good point, fixed.